### PR TITLE
Add `strict_loading` mode to optionally prevent lazy loading

### PR DESCRIPTION
**Pull request by eileencodes**
Add `#strict` to any record to prevent lazy loading of associations.
`strict` will cascade down from the parent record to all the
associations to help you catch any places where you may want to use
`preload` instead of lazy loading.

Co-authored-by: Aaron Patterson <aaron.patterson@gmail.com>

cc/ @rafaelfranca @tenderlove @jhawthorn @matthewd

### CODE REVIEW

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 1]
```diff
@@ -805,10 +805,21 @@ def readonly(value = true)
     end
 
     def readonly!(value = true) # :nodoc:
+      p self.method(:readonly_value).source_location
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/query_methods.rb]: ✂️

File: activerecord/lib/active_record/associations/belongs_to_association.rb [hunk 2]
```diff
@@ -89,6 +89,7 @@ def update_counters_via_scope(klass, foreign_key, by)
         end
 
         def find_target?
+          raise if @owner.readonly?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/associations/belongs_to_association.rb]: Should we raise a specific exception here to tell users what this means? BTW, why `readonly?` should it not be `strict?`

File: activerecord/lib/active_record/associations/preloader/association.rb [hunk 3]
```diff
@@ -124,8 +124,12 @@ def build_scope
             end
 
             scope.merge!(reflection_scope) if reflection.scope
-            scope.merge!(preload_scope) if preload_scope
-            scope
+            scope.merge!(preload_scope) if preload_scope && !preload_scope.empty_scope?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/associations/preloader/association.rb]: I think a two levels if would be easier to understand.

**Review by rafaelfranca — COMMENTED** [review]
I like the idea. Added some comments about the implementation and it would be important to write documentation as well.

**Comment by rafaelfranca** [comment]
Also should we add this option also to the association declaration?

`belongs_to :foo, strict: true`? Right now we kinda of can do that by using a lambda, but maybe it is worth an option.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 1 above)
Comment (eileencodes) [reply activerecord/lib/active_record/relation/query_methods.rb]: Haha oops. This is what I get for opening a PR that I'd been sitting on for months.

File: activerecord/lib/active_record/associations/belongs_to_association.rb (hunk 2 above)
Comment (tenderlove) [reply activerecord/lib/active_record/associations/belongs_to_association.rb]: I think this line can be removed actually (cc @eileencodes).  But yes we should add an exception that is specific to this.

File: activerecord/test/cases/strict_test.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    3.times do
+      AuditLog.create(developer: developer, message: "I am message")
+    end
+
+    dev = Developer.includes(:audit_logs).strict.first
+
+    assert_predicate dev, :strict?
+    assert dev.audit_logs.all?(&:strict?), "Expected all audit logs to be strict"
+  end
+end
```
Comment (dylanahsmith) [inline activerecord/test/cases/strict_test.rb]: This could use more tests for relation methods.  For example, doing `dev.audit_logs.first` in `test_raises_if_strict_and_lazy_loading ` instead of `dev.audit_logs.to_a` will not result in an exception, which I find surprising.  If we are considering the relation itself to be strict, then I would expect that at least relation methods that use the records if they are `loaded?` to raise if the relation hasn't been loaded.

Comment (dylanahsmith) [reply activerecord/test/cases/strict_test.rb]: I think you want the following to handle those methods on the collection proxy

```diff
diff --git a/activerecord/lib/active_record/associations/collection_association.rb b/activerecord/lib/active_record/associations/collection_association.rb
index f1b470f97e..958eba114d 100644
--- a/activerecord/lib/active_record/associations/collection_association.rb
+++ b/activerecord/lib/active_record/associations/collection_association.rb
@@ -306,6 +306,7 @@ def null_scope?
 
       def find_from_target?
         loaded? ||
+          owner.strict? ||
           owner.new_record? ||
           target.any? { |record| record.new_record? || record.changed? }
       end
diff --git a/activerecord/test/cases/strict_test.rb b/activerecord/test/cases/strict_test.rb
index 806810484f..5264521992 100644
--- a/activerecord/test/cases/strict_test.rb
+++ b/activerecord/test/cases/strict_test.rb
@@ -21,6 +21,15 @@ def test_raises_if_strict_and_lazy_loading
     end
   end
 
+  def test_raises_on_unloaded_relation_methods_if_strict
+    dev = Developer.strict.first
+    assert_predicate dev, :strict?
+
+    assert_raises ActiveRecord::StrictViolationError do
+      dev.audit_logs.first
+    end
+  end
+
   def test_preload_audit_logs_are_strict_because_parent_is_strict
     developer = Developer.first
 
```

**Comment by matthewd** [comment]
Is "strict" a sufficiently communicative name? I worry that `User.where(admin: true).strict` reads a little too easily as a stricter form of condition checking, or something. Is e.g. "strict_loading" too wordy?

Sure would be nice if I hadn't left #32136 to wilt for [checks calendar] a year. 😐

**Comment by bogdan** [comment]
Can you please make sure that `strict` is either `false` or `true` but never `nil`. `nil` creates confusion on other relation values like `lock` or `readonly` and I don't think we want to continue that design issue.

**Comment by rafaelfranca** [comment]
`nil` is falsy so there is no problem with returning `nil`. It is following the Rails guidelines for boolean logic.

**Comment by bogdan** [comment]
Problems with ‘nil’ start when you start using ‘*_value’ directly in tests for example or advanced activerecord code. It is just an ambiguity no one wants to deal with.

**Comment by lsylvester** [comment]
I don't think that this currently works with records that are `eager_load`ed in a single query. This test fails for me:

```
  def test_eager_load_audit_logs_are_strict_because_parent_is_strict
    developer = Developer.first

    3.times do
      AuditLog.create(developer: developer, message: "I am message")
    end

    dev = Developer.eager_load(:audit_logs).strict.first

    assert_predicate dev, :strict?
    assert dev.audit_logs.all?(&:strict?), "Expected all audit logs to be strict"
  end

#=> StrictTest#test_preload_audit_logs_are_strict_because_parent_is_strict [/Users/lsylvester/Rails/rails/activerecord/test/cases/strict_test.rb:34]:
    Expected all audit logs to be strict
```

**Comment by louim** [comment]
Hey! This seems like a really useful concept and I would be sad to see this PR disappear in the _stale_ void. 🚀

**Comment by eileencodes** [comment]
Ok we've gotten the extra noted cases working and added some new tests. I cleaned up my messy code. 😬

I like @rafaelfranca's suggestion but that feels like a follow-up PR - maybe even a good first issue for someone to tackle.

I changed the name from `strict` to `strict_loading`.
